Sources/X509/CertificatePrivateKey.swift

```diff
@@ -202,3 +202,81 @@ extension Certificate.PrivateKey {
+    static var pemDiscriminatorForRSAPrivateKey: String { "RSA PRIVATE KEY" }
+    
+    @inlinable
+    static var pemDiscriminatorForSEC1PrivateKey: String { "EC PRIVATE KEY" }
+    
+    @inlinable
+    static var pemDiscriminatorForPKCS8PrivateKey: String { "PRIVATE KEY" }
+    
+    @inlinable
+    public init(pemEncoded: String) throws {
+        try self.init(pemDocument: PEMDocument(pemString: pemEncoded))
+    }
+    
+    @inlinable
+    public init(pemDocument: PEMDocument) throws {
+        switch pemDocument.discriminator {
+        case Self.pemDiscriminatorForRSAPrivateKey:
+            self = try .init(_CryptoExtras._RSA.Signing.PrivateKey.init(derRepresentation: pemDocument.derBytes))
+            
+        case Self.pemDiscriminatorForSEC1PrivateKey:
+            let sec1 = try SEC1PrivateKey(derEncoded: pemDocument.derBytes)
+            self = try .init(ecdsaAlgorithm: sec1.algorithm, rawEncodedPrivateKey: sec1.privateKey.bytes)
+            
+        case Self.pemDiscriminatorForPKCS8PrivateKey:
+            let pkcs8 = try PKCS8PrivateKey(derEncoded: pemDocument.derBytes)
+            switch pkcs8.algorithm {
+            case .ecdsaP256, .ecdsaP384, .ecdsaP521:
+                let sec1 = try SEC1PrivateKey(derEncoded: pkcs8.privateKey.bytes)
+                if let innerAlgorithm = sec1.algorithm, innerAlgorithm != pkcs8.algorithm {
+                    throw ASN1Error.invalidASN1Object(reason: "algorithm missmatch. PKCS#8 is \(pkcs8.algorithm) but inner SEC1 is \(innerAlgorithm)")
+                }
+                self = try .init(ecdsaAlgorithm: pkcs8.algorithm, rawEncodedPrivateKey: sec1.privateKey.bytes)
+                
+            case .rsaKey:
+                self = try .init(_CryptoExtras._RSA.Signing.PrivateKey(derRepresentation: pkcs8.privateKey.bytes))
+            default:
+                throw CertificateError.unsupportedPrivateKey(reason: "unknown algorithm \(pkcs8.algorithm)")
+            }
+            
+        default:
+            throw ASN1Error.invalidPEMDocument(
+                reason: "PEMDocument has incorrect discriminator \(pemDocument.discriminator). Expected \(Self.pemDiscriminatorForPKCS8PrivateKey), \(Self.pemDiscriminatorForSEC1PrivateKey) or \(Self.pemDiscriminatorForRSAPrivateKey) instead"
+            )
+        }
+    }
+    
+    @inlinable
+    init(ecdsaAlgorithm: AlgorithmIdentifier?, rawEncodedPrivateKey: ArraySlice<UInt8>) throws {
+        switch ecdsaAlgorithm {
+        case .some(.ecdsaP256):
+            self = try .init(P256.Signing.PrivateKey(rawRepresentation: rawEncodedPrivateKey))
+        case .some(.ecdsaP384):
+            self = try .init(P384.Signing.PrivateKey(rawRepresentation: rawEncodedPrivateKey))
+        case .some(.ecdsaP521):
+            self = try .init(P521.Signing.PrivateKey(rawRepresentation: rawEncodedPrivateKey))
+        default:
+            throw CertificateError.unsupportedPrivateKey(reason: "unknown algorithm \(ecdsaAlgorithm as Any)")
```

Comment:
Same note on `as Any` here.

Reply:
I generally prefer `String(describing:)` as the fixit suggests

Reply:
oh, good idea. I somehow though it suggested `as Any` but this is probably just me remembering an old warning and not reading the new message 😅 Thanks!